# Data cleaning

## Working with missing data 

In this section, we will discuss missing values also referred to as `NA`, `NaN` or `NULL`  in pandas.

When exploring data, you’ll most likely encounter missing or null values, which symbolically represent non-existent values. 

We can check whether there are missing values in one or more columns of a DataFrame by using `.isna()` or `.isnull()` while `.isna().sum()` or `.isnull().sum()` can be used to get the number of missing values in each of the columns of a DataFrame. 

## Example 1

Import `somalia_women.csv` dataset and count the number of missing values in its columns.

In [1]:
import numpy as np
import pandas as pd

In [2]:
somalia_women = pd.read_csv("datasets/Somalian_women.csv")

In [3]:
somalia_women.head()

,height,weight
0,NaN,64.0
1,63.0,58.0
2,70.0,83.0
3,61.0,NaN
4,64.0,70.0


We check the whether there is a missing value by using `.isna()` or `.isnull()`.

In [4]:
somalia_women.isna()

,height,weight
0,True,False
1,False,False
2,False,False
3,False,True
4,False,False
...,...,...
245,False,False
246,False,False
247,False,True
248,False,False


As you can see, there are some values that are missing in `height` and `weight` of `somalia_women` DataFrame. To get the number of missing values we use `.isna().sum()` or `.isnull().sum().


In [6]:
somalia_women.isna().sum()

height    4
weight    5
dtype: int64

height has $4$ missing values while weight has $5$ missing values.

## Example 2

Import `police_shooting.xlsx` which is in the `datasets` directory and examine its missing values.

Since we have imported all the necessary libraries in Example 1, we just use it directly to import `policeshooting.xlsx` file.

In [7]:
police_shooting = pd.read_excel("datasets/police_shooting.xlsx")

We can then use `.isna().sum()` to check its missing values.

In [8]:
police_shooting.isna().sum()

Unnamed: 0                   0
date                         0
manner_of_death              0
armed                      213
age                        246
gender                       1
race                       537
city                         0
state                        0
signs_of_mental_illness      0
threat_level                 0
flee                       307
body_camera                  0
longitude                  283
latitude                   283
dtype: int64

As you can see, there are missing values in `armed`, `age`, `gender`, `race`, `flee`, `longitude` and `latitude` columns

## Data imputation 

There are two options in dealing with missing values:

1. Get rid of rows or columns with nulls

2. Replace nulls with non-null values which is known as data imputation.

## 1. Get rid of rows or columns with nulls

Every data scientist or analyst regularly face the problem of whether to remove (drop) the missing values (NA, NULL, NAN)  or imputate them. This is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data. Remove nulls is pretty simple by using `.dropna()`. 

---
## Example 1

For example let's remove the missing values in the somalia women dataset. Remember in our `somalia_women` DataFrame, there are some missing values in both height and weight columns.

In [12]:
somalia_women.head()

,height,weight
0,NaN,64.0
1,63.0,58.0
2,70.0,83.0
3,61.0,NaN
4,64.0,70.0


Let's use `.shape` to check the number of rows and columns before removing any missing values.

In [13]:
somalia_women.shape

(250, 2)

In [14]:
somalia_women.isna().sum()

height    4
weight    5
dtype: int64

Now let's drop those missing values with `.dropna()` attribute. This action will delete any row with at least a single NA, Null, or NAN value but it will return a new DataFrame without altering the original one. You could specify `inplace=True` in this method as well.

In [15]:
somalia_women_dropna = somalia_women.dropna()

somalia_women_dropna

,height,weight
1,63.0,58.0
2,70.0,83.0
4,64.0,70.0
5,72.0,67.0
6,59.0,46.0
...,...,...
244,63.0,58.0
245,61.0,52.0
246,63.0,58.0
248,65.0,64.0


In [16]:
somalia_women_dropna.shape

(241, 2)

In [136]:
somalia_women.shape

(250, 2)

You will notice that `dropna()` has reduced the number of rows from $250$ to $241$.

Other than just dropping rows, you can also drop columns with null values by setting axis= "columns" or 1

## Example 2

Import `East-Africa-names.csv` as `east_africa` and remove any column with missing values.

In [17]:
east_africa = pd.read_csv("datasets/East-Africa-names.csv", encoding = "ISO-8859-1")

**Important note**

Sometimes there can be UnicodeDecodeError when reading CSV file in Pandas with Python. You will notice that I used `encoding = "ISO-8859-1"` option to stop this error.

In [18]:
east_africa.head()

,Name,Gender,Origin.1,Origin.2,Origin.3,Origin.4
0,Aadaan,M,Somali,NaN,NaN,NaN
1,Aadam,M,Urdu,Somali,Estonian,NaN
2,Aaden,M,Somali,NaN,NaN,NaN
3,Aamiina,F,Somali,NaN,NaN,NaN
4,Abadir,M,Arabic,Bohairic,Sahidic,NaN


In [19]:
east_africa.isna().sum()

Name           0
Gender         0
Origin.1       0
Origin.2     970
Origin.3    1093
Origin.4    1116
dtype: int64

There are some missing values in Origin.2, Origin.3, Origin.4 of east_Africa DataFrame

Let's check the number of rows and columns before using `.dropna(axis = "columns")` on the `east_africa` DataFrame.

In [20]:
east_africa.shape

(1128, 6)

There are $1128$ rows and $6$ columns. Now let us remove columns with some missing values with `.dropna(axis = "columns")`.

In [21]:
east_africa_dropna = east_africa.dropna(axis = "columns")

east_africa_dropna

,Name,Gender,Origin.1
0,Aadaan,M,Somali
1,Aadam,M,Urdu
2,Aaden,M,Somali
3,Aamiina,F,Somali
4,Abadir,M,Arabic
...,...,...,...
1123,Tufani,M,Swahili
1124,Yordanos,M,Ethiopian
1125,Zaitun,M,Swahili
1126,Zamzam,M,Arabic


In [22]:
east_africa_dropna.shape

(1128, 3)

As you can see, `.dropna(axis = "columns")` has removed all columns with missing values.


## 2. Replace nulls with non-null values which is known as data imputation

Imputation is a conventional technique used to keep valuable data that have missing values. There may be instances where dropping every row with a missing value removes too big a chunk from your dataset, so instead we can impute the missing value with another value, usually the mean or the median of that column.

## Example 1

let's fill the missing values in `height` and `weight` of `somalia_women` DataFrame with their means. The average or mean in each column will be used to replace any missing value found in that column.

In [23]:
somalia_women.head()

,height,weight
0,NaN,64.0
1,63.0,58.0
2,70.0,83.0
3,61.0,NaN
4,64.0,70.0


In [24]:
somalia_women.isna().sum()

height    4
weight    5
dtype: int64

Here's the mean value of `somalia_women` height:

In [25]:
somalia_women["height"].mean() # measured in kg

64.71544715447155

and here is the mean value of `somalia_women's` weight

In [26]:
somalia_women["weight"].mean() # measured in lbs

64.13469387755102

We could also use the code below to achieve the same thing:

In [30]:
somalia_women.mean()

height    64.715447
weight    64.134694
dtype: float64

The mean value of $64.71$ shall be used to replace any `NaN` in height Series of somalia_women DataFrame while the mean value of $64.13$ will also be used to relace any `NaN` in the weight Series.

We can achieve this by using `.fillna()` attribute

In [31]:
somalia_women_imput = somalia_women.fillna(somalia_women.mean())

somalia_women_imput

,height,weight
0,64.715447,64.000000
1,63.000000,58.000000
2,70.000000,83.000000
3,61.000000,64.134694
4,64.000000,70.000000
...,...,...
245,61.000000,52.000000
246,63.000000,58.000000
247,67.000000,64.134694
248,65.000000,64.000000


In [32]:
somalia_women_imput.isna().sum()

height    0
weight    0
dtype: int64

As you can see, all the missing values have been replaced by their mean values.